pkgs/by-name/la/lammps/package.nix

```
@@ -44,9 +45,22 @@
   extraCmakeFlags ? { },
   # Extra `buildInputs` - meant for packages that require more inputs
   extraBuildInputs ? [ ],
+  # If false, the `stdenv` argument will be used no matter what is the state of
+  # extraCmakeFlags.GPU_API or config.cudaSupport . See logic for stdenv' below.
+  overrideStdenv ? true,
+  config ? { },
 }:
 
-stdenv.mkDerivation (finalAttrs: {
+let
+  # Ease adding CUDA support. See:
+  # https://github.com/lammps/lammps/issues/4420
+  stdenv' =
+    if overrideStdenv && (config.cudaSupport || ((extraCmakeFlags.GPU_API or "") == "cuda")) then
```

In case the "cudaSupport" is true do you want to also set the corresponding `cmakeFlags` and `buildInputs` so that lammps gets build automatically with CUDA (i.e. in a similar way we do it in openmpi)?

---

> In case the "cudaSupport" is true do you want to also set the corresponding `cmakeFlags` and `buildInputs` so that lammps gets build automatically with CUDA (i.e. in a similar way we do it in openmpi)?

I think it makes the logic pretty complicated and I'm not sure I want to delve into implementing this. How many details do we want to be taken care of automatically? What do we want to happen in case someone simply sets `config.cudaSupport`? TBH, I don't like this flag at all, as it makes package maintainers' life pretty complicated, especially for complex packages such as this, and since the cuda support for every package is different and users should read the docs of such packages anyway and not simply count on `config.cudaSupport` to do the job...

Here's a list of scenarios I was able to think about that we should handle:

```
  # config.cudaSupport | packages.GPU | extraCmakeFlags.GPU_API || use cudaPackages.backendStdenv? || is valid?
  # ------------------ | ------------ | ----------------------- || ------------------------------- || ---------
  #      false         |    false     |         "opencl"        ||                                 ||
  #      true          |    false     |         "opencl"        ||                                 ||
  #      false         |    true      |         "opencl"        ||                                 ||
  #      true          |    true      |         "opencl"        ||                                 ||
  #      false         |    false     |         "cuda"          ||                                 ||
  #      true          |    false     |         "cuda"          ||                                 ||
  #      false         |    true      |         "cuda"          ||                                 ||
  #      true          |    true      |         "cuda"          ||                                 ||
  #      false         |    false     |         "hip"           ||                                 ||
  #      true          |    false     |         "hip"           ||                                 ||
  #      false         |    true      |         "hip"           ||                                 ||
  #      true          |    true      |         "hip"           ||                                 ||
```

Not only that, there are the cases where `packages.GPU` is not specified, and/or `extraCmakeFlags.GPU_API` is not specified... And, when `opencl` GPU_API is used, you should add the `ocl-icd` package to `buildInputs`, and `cudaPackages.cudatoolkit` when `cuda` GPU_API is used.

I tried for an hour or so to implement a logic that would satisfy everyone and I got lost in the sea of options above. Not only that, I decided to revert the last revision which read the value of `config.cudaSupport`, as I want the expression to be left clean from heuristic decisions that are hard to figure out just from reading the Nix code.

I guess this is the price we pay for the full configurability of `extraCmakeFlags` (and `packages`) and `extraBuildInputs`. I added a `NOTE` comment explaining this decision.

---

I think understand the problem. However, if the config is too complex and if the intention is to configure it externally via `extraCmakeFlags`, I am non sure `overrideStdenv` is even needed here.
I that case you could override it accordingly and have more transparency: 
```
lammps.override {
  stdenv = cudaPackages.backendStdenv;
  extraCmakeFlags.GPU_API = "cuda";
  <+ what else needs to be overriden>
}
```

---

> I think understand the problem. However, if the config is too complex and if the intention is to configure it externally via `extraCmakeFlags`, I am non sure `overrideStdenv` is even needed here. I that case you could override it accordingly and have more transparency:
> 
> ```
> lammps.override {
>   stdenv = cudaPackages.backendStdenv;
>   extraCmakeFlags.GPU_API = "cuda";
>   <+ what else needs to be overriden>
> }
> ```

When you suggested adding an `overrideStdenv`, I thought you wanted to help the following case: A user wishes to build with `-DGPU_API=cuda` (and `-DPKG_GPU=true`), and not use the default `stdenv`, nor `cudaPackages.backendStdenv`, but something like `clangStdenv` (perhaps for testing the build or whatever). Without an `overrideStdenv` flag, this user would have no way to tell the build to use their supplied `stdenv` argument.

The discussion of having an `overrideStdenv` or not doesn't add or subtract from the complexity described in the above table.

---

> `-DGPU_API=cuda`

Our `overrideXXXX` zoo is undeniably a source of infinite pain, but I think for now it's reasonable to say that in these cases "the user is on their own" and they should just use `overrideAttrs` to add `-D` flags manually.

The `if cudaSupport then backendStdenv else stdenv` is an unfortunate pattern since the user now has to `.override {` both `stdenv = ...;` and `cudaPackages = ...; }` to be sure their argument isn't silently ignored, and your `overrideStdenv` solution helps with that but not in a systematic way . Maybe we're better off without it and should build towards changing the `callPackage` mechanism instead

---

> Our `overrideXXXX` zoo is undeniably a source of infinite pain

I agree that it is a zoo, but I wouldn't go as far as saying it is a source of infinite pain :).

> in these cases

Which cases?

> your `overrideStdenv` solution helps with that but not in a systematic way .

Why is your definition for "systematic"? If you ask me, the comment above explains the situation good enough.

> Maybe we're better off without it and should build towards changing the `callPackage` mechanism instead

I have no idea what that means TBH.

---

> Why is your definition for "systematic"? If you ask me, the comment above explains the situation good enough.

We have several places in nixpkgs where we have to come up with an ad hoc name for a `callPackage` parameter, like this `overrideStdenv`, to simultaneously 1) provide a default with conditional logic, 2) ensure there's just one well-defined place to apply the override, 3) prevent the parameter from being substituted by an attribute in the parent scope. There's no way to guess the name of this ad hoc parameter other than by reading the source of the particular derivation. That's "unsystematic".

A systematic solution would be one that separates the "default values" as specified after `?`s in the lambdas from the "default values" as inherited from the parent scope, and from the overrides as passed to `callPackage` and to `.override`. In this case `.override { stdenv = ... ; }` would be just that, instead of `.override { effectiveStdenv = ...; }` or `.override { stdenv = ...; cudaPackages = ...; }`

---

> these cases

Here specifically when the user wants to override the stdenv but still pass the GPU flags. We can try to handle every conceivable combination, but afaiu this is getting complicated, and we'll need to keep maintaining the complexity.

Disclaimer: I may not have familiarized myself with the expression enough to jump to conclusions, I'm not sure if I'm being helpful

---

> Here specifically when the user wants to override the stdenv but still pass the GPU flags. We can try to handle every conceivable combination, but afaiu this is getting complicated, and we'll need to keep maintaining the complexity.

My goal here is only to help Nix users of this package avoid the build error described here:

- https://github.com/lammps/lammps/issues/4420

This build error is barely understandable, and hence it'd be nice to spare user's time by making the usage of both `-DGPU_API="cuda"` and `-DPKG_GPU=ON` CMake flags automatically trigger the usage of the `cudaPackages.backendStdenv`. I want to make this logic as less complicated as possible, and hence introducing `config.cudaSupport` to this logic game is way too much IMO.

Explaining the situation once more and also after the "systematicity" discussion, I think perhaps we can use an assertion that will check whether the `stdenv.cc.version` is lower then or equal to `cudaPackages.backendStdenv.cc.version`. With this logic, we'd spare the `overrideStdenv` argument, while still helping users avoid the experience of the build error mentioned above. What do you think about that?

> Disclaimer: I may not have familiarized myself with the expression enough to jump to conclusions, I'm not sure if I'm being helpful

The expression is not that complicated at the moment, so I think this is not why I personally feel you are not being helpful. I feel you are not being helpful because it takes me too long to understand you.

---

Sorry for the late reply. I do understand your motivation to handle the `stdenv` logic internally. It seems like an acceptable compromise for now.
For now the situation with CUDA is a global mess.

---

> I just thought of another idea that @NixOS/cuda-maintainers might like: Add a setup hook to cudaPackages.cudatoolkit that would prevent it from being compiled with a gcc --version lower then cudaPackages.backendStdenv.cc.version!

Yes, this conversation also made me think in this direction

---

> > I just thought of another idea that @NixOS/cuda-maintainers might like: Add a setup hook to cudaPackages.cudatoolkit that would prevent it from being compiled with a gcc --version lower then cudaPackages.backendStdenv.cc.version!
> 
> Yes, this conversation also made me think in this direction

OK Great, I think this should resolve the whole purpose of this PR. For reference:

- https://github.com/NixOS/nixpkgs/issues/370879

Hence, I'm closing this.